src/sha256t.rs

```diff
@@ -0,0 +1,148 @@
+// Bitcoin Hashes Library
+// Written in 2019 by
+//   The rust-bitcoin developers.
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! # SHA256t (tagged SHA256)
+
+use core::marker::PhantomData;
+
+use sha256;
+use Hash as HashTrait;
+#[allow(unused)]
+use Error;
+
+/// Trait representing a tag that can be used as a context for SHA256t hashes.
+pub trait Tag: Copy + Ord + Default + ::core::hash::Hash {
+    /// Returns a hash engine that is pre-tagged and is ready
+    /// to be used for the data.
+    fn engine() -> sha256::HashEngine;
+}
+
+/// Output of the SHA256t hash function.
+#[derive(Copy, Clone, PartialEq, Eq, Default, PartialOrd, Ord, Hash)]
+pub struct Hash<T: Tag>([u8; 32], PhantomData<T>);
+
+hex_fmt_impl!(Debug, Hash, T:Tag);
+hex_fmt_impl!(Display, Hash, T:Tag);
+hex_fmt_impl!(LowerHex, Hash, T:Tag);
+index_impl!(Hash, T:Tag);
+borrow_slice_impl!(Hash, T:Tag);
+
+impl<T: Tag> HashTrait for Hash<T> {
+    type Engine = sha256::HashEngine;
+    type Inner = [u8; 32];
+
+    fn engine() -> sha256::HashEngine {
+        T::engine()
+    }
+
+    fn from_engine(e: sha256::HashEngine) -> Hash<T> {
+        Hash::from_inner(sha256::Hash::from_engine(e).into_inner())
+    }
+
+    const LEN: usize = 32;
+
+    fn from_slice(sl: &[u8]) -> Result<Hash<T>, Error> {
+        if sl.len() != 32 {
+            Err(Error::InvalidLength(Self::LEN, sl.len()))
+        } else {
+            let mut ret = [0; 32];
+            ret.copy_from_slice(sl);
+            Ok(Hash::from_inner(ret))
+        }
+    }
+
+    // NOTE! If this is changed, please make sure the serde serialization is still correct.
+    const DISPLAY_BACKWARD: bool = true;
+
+    fn into_inner(self) -> Self::Inner {
+        self.0
+    }
+
+    fn from_inner(inner: Self::Inner) -> Self {
+        Hash(inner, PhantomData)
+    }
+}
+
+#[cfg(feature="serde")]
+impl<T: Tag> ::serde::Serialize for Hash<T> {
+    fn serialize<S: ::serde::Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+        use ::hex::ToHex;
+        if s.is_human_readable() {
+            s.serialize_str(&self.to_hex())
+        } else {
+            s.serialize_bytes(&self[..])
+        }
+    }
+}
+
+#[cfg(feature="serde")]
+impl<'de, T: Tag> ::serde::Deserialize<'de> for Hash<T> {
+    fn deserialize<D: ::serde::Deserializer<'de>>(d: D) -> Result<Hash<T>, D::Error> {
+        use ::hex::FromHex;
+
+        if d.is_human_readable() {
+            struct HexVisitor;
+
+            impl<'de> ::serde::de::Visitor<'de> for HexVisitor {
+                type Value = Hash<T>;
```

Comment:
Build with `std-serde` feature fails here: <https://travis-ci.org/rust-bitcoin/bitcoin_hashes/jobs/640164066#L323>

It also happens locally on my machine with nightly version + `std-serde`:

```
error[E0401]: can't use generic parameters from outer function
  --> src/sha256t.rs:98:35
   |
90 | impl<'de, T: Tag> ::serde::Deserialize<'de> for Hash<T> {
   |           - type parameter from outer function
91 |     fn deserialize<D: ::serde::Deserializer<'de>>(d: D) -> Result<Hash<T>, D::Error> {
   |        ----------- try adding a local generic parameter in this method instead
...
98 |                 type Value = Hash<T>;
   |                                   ^ use of generic parameter from outer function
```

I tried different ways of fixing the problem: introducing another generic parameter right on `HexVisitor` impl etc - nothing works here

Reply:
Have fixed it finally: <https://github.com/stevenroose/bitcoin_hashes/pull/3>